src/lightning/pytorch/utilities/testing/_runif.py

```diff
@@ -0,0 +1,197 @@
+
+if module_available("_pytest"):
+    from _pytest.mark import MarkDecorator
+else:
+    MarkDecorator = Any
+
+_SKLEARN_AVAILABLE = RequirementCache("scikit-learn")
+
+
+class _RunIf:
+    """Wrapper around ``pytest.mark.skipif`` with specific conditions.
+
+    @RunIf(min_python="3.6")
+    @pytest.mark.parametrize("arg1", [1, 2.0])
+    def test_wrapper(arg1):
+        assert arg1 > 0.0
+    """
+
+    def __new__(
+        self,
+        min_cuda_gpus: int = 0,
+        min_torch: Optional[str] = None,
+        max_torch: Optional[str] = None,
+        min_python: Optional[str] = None,
+        bf16_cuda: bool = False,
+        tpu: bool = False,
+        ipu: bool = False,
+        hpu: bool = False,
+        mps: Optional[bool] = None,
+        skip_windows: bool = False,
+        standalone: bool = False,
+        deepspeed: bool = False,
+        rich: bool = False,
+        omegaconf: bool = False,
+        psutil: bool = False,
+        sklearn: bool = False,
+        onnx: bool = False,
+        **kwargs: Any,
+    ) -> MarkDecorator:  # not the real return because it would require that pytest is available
+        """
+        Args:
+            min_cuda_gpus: Require this number of gpus and that the ``PL_RUN_CUDA_TESTS=1`` environment variable is set.
+            min_torch: Require that PyTorch is greater or equal than this version.
+            max_torch: Require that PyTorch is less than this version.
+            min_python: Require that Python is greater or equal than this version.
+            bf16_cuda: Require that CUDA device supports bf16.
+            tpu: Require that TPU is available.
+            ipu: Require that IPU is available and that the ``PL_RUN_IPU_TESTS=1`` environment variable is set.
+            hpu: Require that HPU is available.
+            mps: If True: Require that MPS (Apple Silicon) is available,
+                if False: Explicitly Require that MPS is not available
+            skip_windows: Skip for Windows platform.
+            standalone: Mark the test as standalone, our CI will run it in a separate process.
+                This requires that the ``PL_RUN_STANDALONE_TESTS=1`` environment variable is set.
+            deepspeed: Require that microsoft/DeepSpeed is installed.
+            rich: Require that willmcgugan/rich is installed.
+            omegaconf: Require that omry/omegaconf is installed.
+            psutil: Require that psutil is installed.
+            sklearn: Require that scikit-learn is installed.
+            onnx: Require that onnx is installed.
+            **kwargs: Any :class:`pytest.mark.skipif` keyword arguments.
+        """
+        import pytest
+
+        conditions = []
+        reasons = []
+
+        if min_cuda_gpus:
+            conditions.append(num_cuda_devices() < min_cuda_gpus)
+            reasons.append(f"GPUs>={min_cuda_gpus}")
+            # used in conftest.py::pytest_collection_modifyitems
+            kwargs["min_cuda_gpus"] = True
```

Comment:
@carmocca, why we have these strange kwargs used in `pytest_collection_modifyitems`?

Reply:
This is adding a "tag" for filtering when we run with different testing env flags enabled